File: 两个数组的交集II.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# Xiang Wang <ramwin@qq.com>


# 执行用时： 36 ms , 在所有 Python3 提交中击败了 84.72% 的用户
# 内存消耗： 15 MB , 在所有 Python3 提交中击败了 57.57% 的用户

from typing import List
from collections import defaultdict


class Solution:

    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        dict1 = defaultdict(int)
        dict2 = defaultdict(int)
        for i in nums1:
            dict1[i] += 1
        for j in nums2:
            dict2[j] += 1
        result = []
        for k in dict1:
            if k in dict2:
                result += [k] * min(dict1[k], dict2[k])
        return result

File: 两个数组的交集II.cpp
#include <stdio.h>
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;


// 执行用时： 4 ms , 在所有 C++ 提交中击败了 86.79% 的用户
// 内存消耗： 9.7 MB , 在所有 C++ 提交中击败了 81.06% 的用户


class Solution {
  public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
      std::sort(nums1.begin(), nums1.end());
      std::sort(nums2.begin(), nums2.end());
      int index1 = 0;
      int index2 = 0;
      vector<int> result;
      while (index1 < nums1.size() && index2 < nums2.size()) {
        if (nums1[index1] < nums2[index2]) {
          index1++;
        } else if (nums1[index1] > nums2[index2]) {
          index2++;
        } else {
          result.push_back(nums1[index1++]);
          index2++;
        }
      };
      return result;
    }
};
